### PR TITLE
add kubernetes ingress controller (close #25)

### LINKED ISSUE #25: Kubernetes Ingress Controller
## Background

We have developed an ingress controller in EaseMesh, but it is an EaseMesh-dedicated implementation. So we decide to implement a general Kubernetes ingress controller.

## Proposal

The ingress controller consumes Kubernetes Ingress Resources and converts them to an EaseGress configuration which allows the EaseGress to forward and load-balance traffic to Kubernetes pods.

## Design

1. Create a new ingress class

```yaml
apiVersion: networking.k8s.io/v1beta1
kind: IngressClass
metadata:
  name: easegress
spec:
  controller: megaease.com/ingress-controller
```

2. IngressController: An controller in Easegress, also an ingress controller for Kubernetes. It creates a HTTP server for ingress traffic and watches Kubernetes Ingress, Service, Endpoint, Security to translate Kubernetes Ingress rules to Easegress pipelines dynamically.

```yaml
apiVersion: networking.k8s.io/v1beta1
kind: Ingress
metadata:
  name: ingress
spec:
  ingressClassName: easegress
  rules:
  - host: test.example.com
    http:
      paths:
      - path: /
        pathType: Prefix
        backend:
          serviceName: test
          servicePort: 8000
```

the ingress above will be translated to one HTTPServer and one HTTPPipeline

```yaml
kind: HTTPServer
name: http-server-ingress
port: 10080 # Indicated in the installation 
keepAlive: true
keepAliveTimeout: 75s
maxConnection: 10240
rules:
  - host: test.example.com
    paths:
    - path: /
       backend: test
```

```yaml
name: test
kind: HTTPPipeline

flow:
  - filter: proxy
filters:
  - name: proxy
    kind: Proxy
    mainPool:
      servers: # retrieved from k8s
      - url: http://10.0.1.1:19095
      - url: http://10.0.1.2:19095
      loadBalance:
        policy: roundRobin
```

### PR DESCRIPTION
**Pull request by localvar**

### CODE REVIEW

File: doc/ingresscontroller.md [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+      port: 2381
+      nodePort: 2381
+    - name: web
+      protocol: TCP
+      port: 8080
+      nodePort: 80
+  type: NodePort
+```
+
+The service exposes two ports, the `admin` port is for Easegress administration, we exposed it for simplicity here, but you may not want to expose it in your environment for security; the `web` port is to receive external HTTP requests and forward them to the HTTP server in Easegress.
```
Comment (benja-wu) [inline doc/ingresscontroller.md:148]:
```suggestion
The service exposes two ports, the `admin` port is for Easegress administration, we exposed it for simplicity here, for security consideration, we don't recommend you to expose it in your environment; the `web` port is to receive external HTTP requests and forward them to the HTTP server in Easegress.
```

File: pkg/object/ingresscontroller/ingresscontroller.go [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+	Category = supervisor.CategoryBusinessController
+	Kind     = "IngressController"
+
+	defaultIngressClass          = "easegress"
+	defaultIngressControllerName = "megaease.com/ingress-controller"
+	k8sIngressClassAnnotation    = "kubernetes.io/ingress.class"
+)
+
+type (
+	// IngressController implements a k8s ingress controller
```
Comment (benja-wu) [inline pkg/object/ingresscontroller/ingresscontroller.go:41]: `k8s` should be `K8s`

## Reference
1. https://kubernetes.io/docs/concepts/overview/what-is-kubernetes/

File: pkg/object/ingresscontroller/ingresscontroller.go [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+func init() {
+	supervisor.Register(&IngressController{})
+}
+
+// Category returns the category of IngressController.
+func (ic *IngressController) Category() supervisor.ObjectCategory {
+	return Category
+}
+
+// Kind return the kind of IngressController.
```
Comment (benja-wu) [inline pkg/object/ingresscontroller/ingresscontroller.go:73]: `return` should be `returns`

File: pkg/object/ingresscontroller/ingresscontroller.go [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+		case <-time.After(10 * time.Second):
+		}
+	}
+	logger.Infof("successfully watched ingress related resources")
+
+	// process resource update events
+	var err error
+	for {
+		select {
+		case <-ic.stopCh:
```
Comment (benja-wu) [inline pkg/object/ingresscontroller/ingresscontroller.go:166]: Should be the sender's responsibility to close the channel?

## Reference
1. https://tour.golang.org/concurrency/4

File: pkg/object/ingresscontroller/ingresscontroller.go [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+			logger.Errorf("BUG: failed to apply pipeline spec to %s: %v", spec.Name(), err)
+		}
+	}
+	logger.Infof("pipelines updated")
+
+	spec := st.httpServerSpec()
+	_, err = ic.tc.ApplyHTTPServerForSpec(ic.namespace, spec)
+	if err != nil {
+		logger.Errorf("BUG: failed to apply http server spec: %v", err)
+	} else {
```
Comment (benja-wu) [inline pkg/object/ingresscontroller/ingresscontroller.go:228]: Proposal to remove the `Info` log in else branch and above for log cleanliness.

File: pkg/object/ingresscontroller/k8s.go [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+	defaultTimeout = 5 * time.Second
+)
+
+func isResourceEqual(oldObj, newObj interface{}) bool {
+	if oldObj == nil || newObj == nil {
+		return true
+	}
+
+	if oldObj.(metav1.Object).GetResourceVersion() == newObj.(metav1.Object).GetResourceVersion() {
+		return false
```
Comment (benja-wu) [inline pkg/object/ingresscontroller/k8s.go:50]: If `new version` and the `old version` are the same, should return `true` here?

File: pkg/object/ingresscontroller/k8s.go [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+	}
+
+	if ep, ok := oldObj.(*apicorev1.Endpoints); ok {
+		return isEndpointsEqual(ep, newObj.(*apicorev1.Endpoints))
+	}
+
+	return true
+}
+
+func isEndpointsEqual(a, b *apicorev1.Endpoints) bool {
```
Comment (benja-wu) [inline pkg/object/ingresscontroller/k8s.go:60]: According to this function name `isEndpointsEqual`, the logic should be
`if equal returns true, else return false`, right? But it looks the opposite here.

File: pkg/object/ingresscontroller/k8s.go [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+}
+
+type k8sClient struct {
+	namespaces      []string
+	clientset       *kubernetes.Clientset
+	informerFactory informers.SharedInformerFactory
+	eventCh         chan interface{}
+}
+
+// OnDelete is called on Resource Add Events.
```
Comment (benja-wu) [inline pkg/object/ingresscontroller/k8s.go:117]: Inconsistency comment

File: pkg/object/ingresscontroller/k8s.go [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+	// if there's an event already in the channel, discard this one,
+	// this is fine because IngressController always reload everything
+	// when receiving an event, Same for OnUpdate & OnDelete
+	select {
+	case c.eventCh <- obj:
+	default:
+	}
+}
+
+// OnDelete is called on Resource Update Events.
```
Comment (benja-wu) [inline pkg/object/ingresscontroller/k8s.go:128]: Inconsistency comment

File: pkg/object/ingresscontroller/k8s.go [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+}
+
+func (c *k8sClient) getIngresses(ingressClass string) []*apinetv1.Ingress {
+	var result []*apinetv1.Ingress
+
+	lister := c.informerFactory.Networking().V1().Ingresses().Lister()
+	for _, ns := range c.namespaces {
+		list, err := lister.Ingresses(ns).List(labels.Everything())
+		if err != nil {
+			logger.Errorf("Failed to get ingresses from namespace %s: %v", ns, err)
```
Comment (benja-wu) [inline pkg/object/ingresscontroller/k8s.go:238]: Proposal to add `continue` here.

File: pkg/object/ingresscontroller/translator.go [hunk 11]
```diff
@@ -0,0 +1,356 @@
...
+
+func (st *specTranslator) pipelineSpecs() map[string]*supervisor.Spec {
+	return st.pipelines
+}
+
+func generatePipelineSpec(name string, endpoints []string) (*supervisor.Spec, error) {
+	var buf bytes.Buffer
+	fmt.Fprintf(&buf, "name: %s\n", name)
+	buf.WriteString(`kind: HTTPPipeline
+flow:
```
Comment (benja-wu) [inline pkg/object/ingresscontroller/translator.go:68]: According to @xxx7xxxx 's advice in MeshController and FaaSController, proposal to use structure builder[1]

## Reference
1. https://github.com/megaease/easegress/blob/main/pkg/object/function/worker/ingress.go#L62

File: pkg/object/ingresscontroller/ingresscontroller.go (hunk 4 above)
Comment (localvar) [reply pkg/object/ingresscontroller/ingresscontroller.go:166]: `ic->stopCh` and `stopCh` are two channels.

Comment (benja-wu) [reply pkg/object/ingresscontroller/ingresscontroller.go:166]: If so, propose to move the stop channel into the `k8sclient` structure as its `watch` API's return value.  The `IngressController` only needs to receive this channel's closing message.

**Comment by localvar** [comment]
updated according to comments.

**Review by benja-wu — APPROVED** [review]
LGTM

**Comment by Loyalsoldier** [comment]
Some PRs merged before didn't pass the `check_dep` test.

**Comment by benja-wu** [comment]
> Some PRs merged before didn't pass the `check_dep` test.

Thanks for reminding. 
* Ubuntu test failed: mostly it's about `go mod tidy` changed causing `git diff ` checking failed, that can be solved by updating `go.mod` or `go.sum` with `make test` in the local environment. @localvar 
* MacOS test failed: we already opening a PR for fixing this problem.( #107 )

File: pkg/object/ingresscontroller/ingresscontroller.go [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+		ObjectStatus: nil,
+	}
+}
+
+// Close closes IngressController.
+func (ic *IngressController) Close() {
+	close(ic.stopCh)
+	ic.wg.Wait()
+
+	for _, entity := range ic.tc.ListHTTPServers(ic.namespace) {
```
Comment (xxx7xxxx) [inline pkg/object/ingresscontroller/ingresscontroller.go:193]: Use `tc.Clean`

Comment (localvar) [reply pkg/object/ingresscontroller/ingresscontroller.go]: updated

**Review by xxx7xxxx — APPROVED** [review]

**Pull request merged**